File: triton-soccer-ai/src/main/java/com/triton/module/interface_module/SimulatorRobotControlInterface.java
package com.triton.module.interface_module;

import com.rabbitmq.client.Delivery;
import com.triton.config.NetworkConfig;
import com.triton.constant.RuntimeConstants;
import com.triton.module.Module;
import com.triton.networking.UDP_Client;

import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.net.DatagramPacket;
import java.util.concurrent.TimeoutException;

import static com.triton.config.ConfigPath.NETWORK_CONFIG;
import static com.triton.config.ConfigReader.readConfig;
import static com.triton.messaging.Exchange.ROBOT_CONTROL;
import static com.triton.messaging.SimpleSerialize.simpleDeserialize;
import static proto.simulation.SslSimulationRobotControl.RobotControl;
import static proto.simulation.SslSimulationRobotFeedback.RobotControlResponse;

public class SimulatorRobotControlInterface extends Module {
    private NetworkConfig networkConfig;

    private UDP_Client client;

    public SimulatorRobotControlInterface() throws IOException, TimeoutException {
        super();
        setupClient();
        declareExchanges();
    }

    @Override
    protected void loadConfig() throws IOException {
        super.loadConfig();
        networkConfig = (NetworkConfig) readConfig(NETWORK_CONFIG);
    }

    private void setupClient() throws IOException {
        String allyControlAddress;
        int allyControlPort;

        switch (RuntimeConstants.team) {
            case BLUE -> {
                allyControlAddress = networkConfig.getSimulationRobotControlBlueAddress();
                allyControlPort = networkConfig.getSimulationRobotControlBluePort();
            }
            case YELLOW -> {
                allyControlAddress = networkConfig.getSimulationRobotControlYellowAddress();
                allyControlPort = networkConfig.getSimulationRobotControlYellowPort();
            }
            default -> throw new IllegalStateException("Unexpected value: " + RuntimeConstants.team);
        }

        client = new UDP_Client(allyControlAddress, allyControlPort, this::callbackRobotControlResponse);
        client.start();
    }

    @Override
    protected void declareExchanges() throws IOException, TimeoutException {
        super.declareExchanges();
        declareConsume(ROBOT_CONTROL, this::callbackRobotControl);
    }

    private void callbackRobotControl(String s, Delivery delivery) {
        RobotControl robotControl;
        try {
            robotControl = (RobotControl) simpleDeserialize(delivery.getBody());
        } catch (IOException | ClassNotFoundException e) {
            e.printStackTrace();
            return;
        }

        client.send(robotControl.toByteArray());
    }

    private void callbackRobotControlResponse(DatagramPacket packet) {
        try {
            RobotControlResponse response = parsePacket(packet);
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    private RobotControlResponse parsePacket(DatagramPacket packet) throws IOException {
        ByteArrayInputStream stream = new ByteArrayInputStream(packet.getData(),
                packet.getOffset(),
                packet.getLength());
        RobotControlResponse response = RobotControlResponse.parseFrom(stream);
        stream.close();
        return response;
    }
}

File: triton-bot/src/main/python/module/interface_module/simulator_robot_control_interface.py
from pickle import loads

from config.config_path import ConfigPath
from config.config_reader import read_config
from generated_sources.proto.ssl_simulation_robot_control_pb2 import (
    RobotCommand, RobotControl)
from generated_sources.proto.ssl_simulation_robot_feedback_pb2 import \
    RobotControlResponse
from constant.runtime_constants import RuntimeConstants
from constant.team import Team
from messaging.exchange import Exchange
from module.module import Module
from networking.udp_client import UDP_Client


class SimulatorRobotControlInterface(Module):
    def __init__(self):
        super().__init__()
        self.last_time_stamp = 0

    def load_config(self):
        super().load_config()
        self.network_config = read_config(
            ConfigPath.NETWORK_CONFIG)

    def prepare(self):
        super().prepare()
        self.setup_client()

    def declare_consumes(self):
        super().declare_consumes()
        self.declare_consume(exchange=Exchange.TB_LOCAL_COMMAND,
                             callback=self.callback_local_command)

    def run(self):
        super().run()
        self.consume()

    def setup_client(self):
        if (RuntimeConstants.team == Team.YELLOW):
            ally_control_address = self.network_config['simulationRobotControlAddressYellow']
            ally_control_port = self.network_config['simulationRobotControlPortYellow']
        else:
            ally_control_address = self.network_config['simulationRobotControlAddressBlue']
            ally_control_port = self.network_config['simulationRobotControlPortBlue']

        self.client = UDP_Client(
            server_address=ally_control_address,
            server_port=ally_control_port,
            callback=self.callback_robot_control_response)

        self.client.start()

    def callback_local_command(self, ch, method, properties, body):
        local_command = RobotCommand()
        local_command.ParseFromString(body)
        print(local_command)
        robot_control = RobotControl()
        robot_control.robot_commands.append(local_command)
        self.client.add_send(robot_control.SerializeToString())

    def callback_robot_control_response(self, bytes):
        response = RobotControlResponse()
        response.ParseFromString(bytes)
        for feedback in response.feedback:
            if (feedback.id != RuntimeConstants.id):
                continue
            self.publish(exchange=Exchange.TB_FEEDBACK, object=feedback)